src/custom-element.js

```diff
@@ -1210,6 +1210,17 @@ function createBaseCustomElementClass(win) {
     }
 
     /**
+     * Called when an attribute's value changes.
+     * Only called for observedAttributes or from amp-bind.
+     * @param {!string} name
+     * @param {?string} oldValue
+     * @param {?string} newValue
+     */
+    attributeChangedCallback(name, oldValue, newValue) {
```

Comment:
I think it is a better design to not align ourselves with the custom elements spec and instead call APIs in various lifecycle of the state changes.

We might need attributeChangedCallback, but I think it should only be used to translate those changes into state changes